src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -809,63 +809,199 @@ ErrorIfInsertPartitionColumnDoesNotMatchSelect(Query *query, RangeTblEntry *inse
+							break;
+						}
+
+						default:
+						{
+							exprDescription = "a function call";
+							break;
+						}
+					}
+					break;
+				}
+
+				case T_Aggref:
+				{
+					exprDescription = "an aggregation";
+					break;
+				}
+
+				case T_CaseExpr:
+				{
+					exprDescription = "a case expression";
+					break;
+				}
+
+				case T_CoalesceExpr:
+				{
+					exprDescription = "a coalesce expression";
+					break;
+				}
+
+				case T_RowExpr:
+				{
+					exprDescription = "a row expression";
+					break;
+				}
+
+				case T_MinMaxExpr:
+				{
+					exprDescription = "a min/max expression";
+					break;
+				}
+
+				case T_CoerceViaIO:
+				{
+					exprDescription = "an explicit coercion";
+					break;
+				}
+
+				default:
+				{
+					exprDescription = "an expression that is not a plain column";
+					break;
+				}
 			}
 
-			partitionColumnsMatch = true;
-			*selectPartitionColumnTableId = subqeryTargetEntry->resorigtbl;
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot plan distributed INSERT INTO ... SELECT query "
+								   "because of partition column restrictions"),
+							errdetail(errorDetailTemplate, exprDescription),
+							errhint("Ensure the target table's partition column has a "
+									"corresponding simple reference to a distributed "
+									"table's partition column in the subquery.")));
+		}
 
-			break;
+		/*
+		 * Insert target expression could only be non-var if the select target
+		 * entry does not have the same type (i.e., target column requires casting).
+		 */
+		if (!IsA(targetEntry->expr, Var))
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot plan distributed INSERT INTO ... SELECT query "
+								   "because partition column types do not match"),
+							errdetail(
+								"The data type of the target table's partition column "
+								"should exactly match the data type of the "
+								"corresponding target list entry in the subquery.")));
+		}
+
+		/* finally, check that the select target column is a partition column */
+		if (!IsPartitionColumnRecursive(selectTargetExpr, subquery))
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot plan distributed INSERT INTO ... SELECT query "
+								   "because of mismatched partition columns"),
+							errdetail(
+								"The target table's partition column "
+								"should correspond to a partition column "
+								"in the subquery.")));
 		}
+
+		/* set this for the caller function */
+		*selectPartitionColumnTableId = subqueryTargetEntry->resorigtbl;
+
+		break;
 	}
 
-	if (!partitionColumnsMatch)
+
+	if (!targetTableHasPartitionColumn)
 	{
 		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-						errmsg("SELECT query should return bare partition column on "
-							   "the same ordinal position as the INSERT's partition "
-							   "column")));
+						errmsg("cannot plan distributed INSERT INTO ... SELECT query"),
```

Comment:
This is the only place where the `errmsg` does not have `because...`. I'm thinking whether to update this or not. What do you think?